portfolio/src/test/java/com/google/sps/TimeRangeGroupTest.java

```diff
@@ -0,0 +1,669 @@
+    Instant timeRangeToAddEnd = timeRangeTwoEnd.plusSeconds(1000);
+    TimeRange timeRangeToAdd = TimeRange.fromStartEnd(timeRangeToAddStart, timeRangeToAddEnd);
+
+    // Constructs the expected merged time range for combining B and C.
+    Instant timeRangeCombinedStart = timeRangeTwoStart;
+    Instant timeRangeCombinedEnd = timeRangeToAddEnd;
+    TimeRange timeRangeNew = TimeRange.fromStartEnd(timeRangeCombinedStart, timeRangeCombinedEnd);
+    List<TimeRange> expected = Arrays.asList(timeRangeOne, timeRangeNew);
+
+    timeRangeGroup.addTimeRange(timeRangeToAdd);
+    Iterator<TimeRange> actualIterator = timeRangeGroup.iterator();
+
+    List<TimeRange> actual = new LinkedList();
+    while (actualIterator.hasNext()) {
+      actual.add(actualIterator.next());
+    }
+
+    Collections.sort(actual, TimeRange.SORT_BY_TIME_RANGE_DURATION_ASCENDING_THEN_START_TIME);
+    Assert.assertEquals(expected, actual);
+  }
+
+  /** Tests for adding a time range that entirely contains an existing one. */
+  @Test
+  public void testAddSupersetRange() {
+    // Time Ranges:      |-A-|
+    // To add:       |-----B-----|
+    // Result:       |-----------|
+    Instant timeRangeOneStart = Instant.now();
+    Instant timeRangeOneEnd = timeRangeOneStart.plusSeconds(3000);
+    Instant timeRangeTwoStart = timeRangeOneStart.plusSeconds(500);
+    Instant timeRangeTwoEnd = timeRangeTwoStart.plusSeconds(500);
+
+    TimeRange timeRangeOne = TimeRange.fromStartEnd(timeRangeOneStart, timeRangeOneEnd);
+    TimeRange timeRangeTwo = TimeRange.fromStartEnd(timeRangeTwoStart, timeRangeTwoEnd);
+
+    List<TimeRange> originalTimeRanges = Arrays.asList(timeRangeTwo);
+    for (TimeRange timeRange : originalTimeRanges) {
+      timeRangeGroup.addTimeRange(timeRange);
+    }
+    List<TimeRange> expected = Arrays.asList(timeRangeOne);
+
+    timeRangeGroup.addTimeRange(timeRangeOne);
+    Iterator<TimeRange> actualIterator = timeRangeGroup.iterator();
+
+    List<TimeRange> actual = new LinkedList();
+    while (actualIterator.hasNext()) {
+      actual.add(actualIterator.next());
+    }
+
+    Collections.sort(actual, TimeRange.SORT_BY_TIME_RANGE_DURATION_ASCENDING_THEN_START_TIME);
+    Assert.assertEquals(expected, actual);
+  }
+
+  /** Tests for adding a time range that overlaps another existing one. */
+  @Test
+  public void testAddExistingRange() {
+    // Time Ranges:  |-----A-----|
+    // To add:       |-----B-----|
+    // Result:       |-----------|
+    Instant timeRangeOneStart = Instant.now();
+    Instant timeRangeOneEnd = timeRangeOneStart.plusSeconds(1000);
+
+    TimeRange timeRangeOne = TimeRange.fromStartEnd(timeRangeOneStart, timeRangeOneEnd);
+
+    List<TimeRange> originalTimeRanges = Arrays.asList(timeRangeOne);
+    for (TimeRange timeRange : originalTimeRanges) {
+      timeRangeGroup.addTimeRange(timeRange);
+    }
+    List<TimeRange> expected = Arrays.asList(timeRangeOne);
+
+    timeRangeGroup.addTimeRange(timeRangeOne);
```

Comment:
Hmmm what if you try to construct a new TimeRange instance called `timeRangeTwo`, and initialize this new time range with the exact start and end times of timeRangeOne? I think in this case the test passes for array list time range group, at least it did when I tried it on my own.

Reply:
Seems like I just had outdated code, ArrayListTimeRangeGroup passes this test just fine now.